### ISSUE TITLE
How can I save the checkpoint by iteration, also in Epochs too

### ISSUE THREAD

**Issue by yellowjs0304** [open-mmlab/mmocr#670]
Hi, I wonder How can i save the checkpoint by iteration and Epochs. It takes really long time for me around 5 days for 1 epoch... So, I wanna see the model's evaluation in every 10000 iterations. It seems like modifying default_runtime.py(config) or the other config files, but I can' t see the right way. Could you please share the info?

I modified two files as below, and it returns save the model every iteration, but the evaluation seems doesn't work.
![image](https://user-images.githubusercontent.com/30281608/145913639-e708e3b5-e128-49ad-b870-c70507689c76.png)

I wonder this is the right way?
```
[default_runtime.py]
checkpoint_config = dict(interval=100, by_epoch=False, max_keep_ckpts = 3)
# yapf:disable
log_config = dict(
    interval=10,
    hooks=[
        dict(type='TextLoggerHook'),
        dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]

---
[configs/textrecog/satrn/satrn.py]

# optimizer
optimizer = dict(type='Adam', lr=3e-4)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy='step', step=[3, 4])
total_epochs = 5

data = dict(
    samples_per_gpu=32,
    workers_per_gpu=4,
    val_dataloader=dict(samples_per_gpu=4),
    test_dataloader=dict(samples_per_gpu=4),
    train=dict(
        type='UniformConcatDataset',
        datasets=train_list,
        pipeline=train_pipeline),
    val=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline),
    test=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline))

evaluation = dict(interval=100, metric='acc', by_epoch=False)
```

**Comment by gaotongxiao** [comment]
There's no way to save checkpoints on both iter and epoch base for now. 

Your config looks fine overall. The only discrepancy I could find was that you want to evaluate every 10000 iterations, but the last line of your config shows you were evaluating the model every 100 iterations. I can't tell if your evaluation works appropriately since your screenshot only shows the log of the first 80 iterations.

**Comment by yellowjs0304** [comment]
10000 iter is just typo. I set same number both of them. And the screen-shot which has 80 iters is because i loaded pre-trained model, so model already has initial iteration. I think that's the reason why checkpoint saved in 80 iters.

**Comment by yellowjs0304** [comment]
I tested one experiment that didn't load pre-trained model(=not to resume) . Saving model in every N iters is great, But It still doesn't evaluate..

![image](https://user-images.githubusercontent.com/30281608/145948020-16011e45-3ad0-4a91-b7c1-4f2cf85a5e72.png)

Am I need another file to modified?? Please tell me what i missed.

**Comment by gaotongxiao** [comment]
You'll need to specify the runner type as `IterBasedRunner`, otherwise it uses `EpochBasedRunner` by default and will never trigger iter-based evaluation. 
```python
# total_epochs = 5
evaluation = dict(interval=100, metric='acc', by_epoch=False)
runner = {'type': 'IterBasedRunner', 'max_iters': 29476400}
checkpoint_config = dict(interval=100)
```
It's a bit confusing and looks like a design flaw.

**Comment by yellowjs0304** [comment]
I understand. Thank you

**Comment by yellowjs0304** [comment]
Hi, I reopened this issue. I did Iteration training, and there's an error when evaluation 
I think this is related with below issue. 

I modified(added 3 lines bottom of code) the config file(textrecog/satrn/satrn.py)
```
train_list = {{_base_.train_list}}
test_list = {{_base_.test_list}}

train_pipeline = {{_base_.train_pipeline}}
test_pipeline = {{_base_.test_pipeline}}

label_convertor = dict(
    type='AttnConvertor', dict_type='KOR', with_unknown=True)

model = dict(
    type='SATRN',
    backbone=dict(type='ShallowCNN', input_channels=3, hidden_dim=512),
    encoder=dict(
        type='SatrnEncoder',
        n_layers=12,
        n_head=8,
        d_k=512 // 8,
        d_v=512 // 8,
        d_model=512,
        n_position=100,
        d_inner=512 * 4,
        dropout=0.1),
    decoder=dict(
        type='NRTRDecoder',
        n_layers=6,
        d_embedding=512,
        n_head=8,
        d_model=512,
        d_inner=512 * 4,
        d_k=512 // 8,
        d_v=512 // 8),
    loss=dict(type='TFLoss'),
    label_convertor=label_convertor,
    max_seq_len=25)

# optimizer
optimizer = dict(type='Adam', lr=3e-4)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy='step', step=[3, 4])
#total_epochs = 5

data = dict(
    samples_per_gpu=64,
    workers_per_gpu=4,
    val_dataloader=dict(samples_per_gpu=32),
    test_dataloader=dict(samples_per_gpu=32),
    train=dict(
        type='UniformConcatDataset',
        datasets=train_list,
        pipeline=train_pipeline),
    val=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline),
    test=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline))

# THIS LINE - for Iteration Learning
evaluation = dict(interval=1000, metric='acc', by_epoch = False)
runner = {'type': 'IterBasedRunner', 'max_iters': 37569400}
checkpoint_config = dict(interval=1000, by_epoch=False, max_keep_ckpts = 3)
```
![image](https://user-images.githubusercontent.com/30281608/147065369-afb65fcd-8ee9-48de-a33a-4926c21a2eeb.png)

https://github.com/open-mmlab/mmdetection/pull/5882

How can i fix this issue?

**Comment by gaotongxiao** [comment]
I think you have already found the solution yourself - you just need to modify MMOCR's `train_detector` in `mmocr/apis/train.py`  in the same way as that PR. If you find this patch really works, you are also welcome to raise a PR to help us fix this issue :)

**Comment by yellowjs0304** [comment]
okay, thank you for answering.

Also, I have another question about learning_rate.

I want decrease the learning rate along every iters. 
I followed same iter, policy as explained above.
```
# optimizer
optimizer = dict(type='Adam', lr=3e-4)
optimizer_config = dict(grad_clip=None)
# learning policy
lr_config = dict(policy='step', step=[3, 4])
#total_epochs = 5
```
I thought it could be change automatically, because i modified the runner by iteration.
But, the learning rate still same value in 5000 iters.
![image](https://user-images.githubusercontent.com/30281608/147170288-9bb531ed-7ba1-48a8-be3d-00ce4699a709.png)

I think it needs another part too.. What can i do?

**Comment by gaotongxiao** [comment]
Passing a list of steps to `step` policy means that lr only decays at your specific steps. You should pass an integer X instead so that lr would decay every X steps. Ref:
https://github.com/open-mmlab/mmcv/blob/690a77fa6f6e3f301b77204f985669a237042b4e/mmcv/runner/hooks/lr_updater.py#L167

**Comment by yellowjs0304** [comment]
Thank you so much. 
i also found there's an error with my command "--resume-from " This option take over the lr scheduler, optimizers from the loaded weights, so It cannot apply the modified options.

I modified this by --load-from, and now it is start with epoch 0, apply well lr-scheduler that i modified.

Thank you again :)
